File: compute/officejs/src/application.js
(function (global) {
  "use strict";

  var Excel = global.Excel;
  var OfficeExtension = global.OfficeExtension;
  var ClientObject = OfficeExtension.ClientObject;
  var officeJs = global.__mogOfficeJs;

  function error(code, message) {
    return new OfficeExtension.Error({ code: code, message: message });
  }

  function propertyNotLoaded(name) {
    return error(
      "PropertyNotLoaded",
      "The property '" +
        name +
        "' is not available. Before reading the property's value, call the load method on the containing object and call \"context.sync()\" on the associated request context."
    );
  }

  function requirePropertyObject(source, typeName) {
    if (source == null || typeof source !== "object") {
      throw error("InvalidArgument", typeName + ".set requires a property object");
    }
  }

  function assertSameContext(source, target) {
    if (source.context !== target.context) {
      throw error(
        "InvalidRequestContext",
        "The object belongs to a different request context."
      );
    }
  }

  function loadedValue(object, name) {
    if (!object._loaded[name]) throw propertyNotLoaded(name);
    return object["_" + name];
  }

  function isClientObject(value) {
    return value instanceof ClientObject;
  }

  function queueSet(object, property, value, ensureBinding) {
    if (ensureBinding) ensureBinding.call(object);
    object["_" + property] = value;
    object._loaded[property] = true;
    object.context._queue.push({
      op: "set",
      id: object._id,
      property: property,
      value: value,
    });
  }

  function enumValue(value, property, allowed) {
    if (allowed.indexOf(value) < 0) {
      throw error(
        "InvalidArgument",
        property + " must be one of: " + allowed.join(", ")
      );
    }
    return value;
  }

  function readOnlyProperties(typeName) {
    return function (name, properties, throwOnReadOnly) {
      if (properties[name] === undefined) return true;
      if (throwOnReadOnly) {
        throw error("InvalidArgument", typeName + "." + name + " is read-only");
      }
      return true;
    };
  }

  function sourceObject(target, source, typeName) {
    requirePropertyObject(source, typeName);
    if (!isClientObject(source)) {
      return { properties: source };
    }
    assertSameContext(source, target);
    if (Object.getPrototypeOf(source) !== Object.getPrototypeOf(target)) {
      throw error(
        "InvalidArgument",
        "The object passed to " + typeName + ".set must have the same type."
      );
    }
    return { properties: source.toJSON() };
  }

  function setProperties(target, source, typeName, writable, readOnly, options) {
    var normalized = sourceObject(target, source, typeName);
    var properties = normalized.properties;
    var throwOnReadOnly = !(options && options.throwOnReadOnly === false);

    Object.keys(properties).forEach(function (name) {
      if (writable.indexOf(name) >= 0) return;
      if (readOnly.indexOf(name) >= 0) {
        readOnlyProperties(typeName)(
          name,
          properties,
          throwOnReadOnly
        );
        return;
      }
      throw error("InvalidArgument", "Unknown " + typeName + " property: " + name);
    });

    writable.forEach(function (name) {
      if (Object.prototype.hasOwnProperty.call(properties, name)) {
        if (properties[name] !== undefined) target[name] = properties[name];
      }
    });
    return properties;
  }

  function Application(context, workbook) {
    ClientObject.call(this, context);
    this._workbook = workbook || null;
    // These are the scalar members that this headless host can answer from
    // the compute engine. Unsupported desktop/locale members remain explicit
    // descriptors below and are never silently fabricated by load().
    this._scalarProperties = ["calculationMode", "calculationState"];
    this._navigationProperties = ["iterativeCalculation"];
    this._bindingQueued = false;
  }
  Application.prototype = Object.create(ClientObject.prototype);
  Application.prototype.constructor = Application;

  Application.prototype._ensureBinding = function () {
    if (this._bindingQueued) return;
    this.context._queue.push({ op: "applicationBind", id: this._id });
    this._bindingQueued = true;
  };

  Application.prototype.load = function (props) {
    this._ensureBinding();
    return ClientObject.prototype.load.call(this, props);
  };

  function applicationScalar(name, setter) {
    Object.defineProperty(Application.prototype, name, {
      configurable: true,
      get: function () {
        return loadedValue(this, name);
      },
      set: setter,
    });
  }

  applicationScalar("calculationMode", function (value) {
    value = enumValue(value, "Application.calculationMode", [
      "Automatic",
      "AutomaticExceptTables",
      "Manual",
    ]);
    queueSet(this, "calculationMode", value, this._ensureBinding);
  });
  applicationScalar("calculationState");

  // These members exist in the pinned declarations, but require host
  // capabilities that this compute-backed runtime does not provide. Keeping
  // a normal unloaded getter makes an explicit load fail at sync and avoids a
  // misleading local value.
  [
    "calculationEngineVersion",
    "decimalSeparator",
    "thousandsSeparator",
    "useSystemSeparators",
  ].forEach(function (name) {
    applicationScalar(name);
  });

  function IterativeCalculation(context, application) {
    ClientObject.call(this, context);
    this._application = application;
    this._scalarProperties = ["enabled", "maxChange", "maxIteration"];
    this._bindingQueued = false;
  }
  IterativeCalculation.prototype = Object.create(ClientObject.prototype);
  IterativeCalculation.prototype.constructor = IterativeCalculation;

  IterativeCalculation.prototype._ensureBinding = function () {
    if (this._bindingQueued) return;
    this._application._ensureBinding();
    this.context._queue.push({
      op: "iterativeCalculationBind",
      id: this._id,
      applicationId: this._application._id,
    });
    this._bindingQueued = true;
  };

  IterativeCalculation.prototype.load = function (props) {
    this._ensureBinding();
    return ClientObject.prototype.load.call(this, props);
  };

  ["enabled", "maxChange", "maxIteration"].forEach(function (name) {
    Object.defineProperty(IterativeCalculation.prototype, name, {
      configurable: true,
      get: function () {
        return loadedValue(this, name);
      },
      set: function (value) {
        if (name === "enabled" && typeof value !== "boolean") {
          throw error("InvalidArgument", "IterativeCalculation.enabled must be a boolean");
        }
        if (
          name === "maxChange" &&
          (typeof value !== "number" || !isFinite(value) || value < 0)
        ) {
          throw error(
            "InvalidArgument",
            "IterativeCalculation.maxChange must be a finite non-negative number"
          );
        }
        if (
          name === "maxIteration" &&
          (typeof value !== "number" ||
            !isFinite(value) ||
            Math.floor(value) !== value ||
            value < 0)
        ) {
          throw error(
            "InvalidArgument",
            "IterativeCalculation.maxIteration must be a non-negative integer"
          );
        }
        queueSet(this, name, value, this._ensureBinding);
      },
    });
  });

  IterativeCalculation.prototype.set = function (source) {
    setProperties(
      this,
      source,
      "IterativeCalculation",
      ["enabled", "maxChange", "maxIteration"],
      [],
      undefined
    );
  };

  IterativeCalculation.prototype.toJSON = function () {
    var data = {};
    this._scalarProperties.forEach(function (name) {
      if (this._loaded[name]) data[name] = this["_" + name];
    }, this);
    return data;
  };

  Object.defineProperty(Application.prototype, "iterativeCalculation", {
    configurable: true,
    get: function () {
      this._ensureBinding();
      if (!this._iterativeCalculation) {
        this._iterativeCalculation = new IterativeCalculation(this.context, this);
      }
      return this._iterativeCalculation;
    },
  });

  Application.prototype.set = function (source, options) {
    var normalized = sourceObject(this, source, "Application");
    var properties = normalized.properties;
    var throwOnReadOnly = !(options && options.throwOnReadOnly === false);
    Object.keys(properties).forEach(function (name) {
      if (name === "calculationMode" || name === "iterativeCalculation") return;
      if (
        [
          "activeWindow",
          "cultureInfo",
          "calculationEngineVersion",
          "calculationState",
          "decimalSeparator",
          "thousandsSeparator",
          "useSystemSeparators",
          "windows",
        ].indexOf(name) >= 0
      ) {
        readOnlyProperties("Application")(name, properties, throwOnReadOnly);
        return;
      }
      throw error("InvalidArgument", "Unknown Application property: " + name);
    });
    if (
      Object.prototype.hasOwnProperty.call(properties, "calculationMode") &&
      properties.calculationMode !== undefined
    ) {
      this.calculationMode = properties.calculationMode;
    }
    if (Object.prototype.hasOwnProperty.call(properties, "iterativeCalculation")) {
      var child = this.iterativeCalculation;
      if (properties.iterativeCalculation && typeof properties.iterativeCalculation === "object") {
        child.set(properties.iterativeCalculation);
      }
    }
  };

  Application.prototype.calculate = function (calculationType) {
    calculationType = enumValue(calculationType, "Application.calculate calculationType", [
      "Recalculate",
      "Full",
      "FullRebuild",
    ]);
    this._ensureBinding();
    this.context._queue.push({
      op: "applicationCalculate",
      id: this._id,
      calculationType: calculationType,
    });
  };

  Application.prototype.toJSON = function () {
    var data = {};
    this._scalarProperties.forEach(function (name) {
      if (this._loaded[name]) data[name] = this["_" + name];
    }, this);
    if (this._iterativeCalculation) {
      data.iterativeCalculation = this._iterativeCalculation.toJSON();
    }
    return data;
  };

  function ensureWorkbookBinding() {
    if (this._workbookBindingQueued) return;
    this.context._queue.push({ op: "workbookBind", id: this._id });
    this._workbookBindingQueued = true;
  }

  // Workbook is created by bootstrap. The extension adds the one scalar
  // member owned by this family and a binding operation for generic load/set.
  officeJs.addScalarProperties(Excel.Workbook.prototype, [
    "usePrecisionAsDisplayed",
  ]);
  officeJs.addNavigationProperties(Excel.Workbook.prototype, ["application"]);

  Object.defineProperty(Excel.Workbook.prototype, "application", {
    configurable: true,
    get: function () {
      if (!this._application) this._application = new Application(this.context, this);
      return this._application;
    },
  });

  var workbookLoad = Excel.Workbook.prototype.load;
  Excel.Workbook.prototype.load = function (props) {
    ensureWorkbookBinding.call(this);
    return workbookLoad.call(this, props);
  };

  Object.defineProperty(Excel.Workbook.prototype, "usePrecisionAsDisplayed", {
    configurable: true,
    get: function () {
      return loadedValue(this, "usePrecisionAsDisplayed");
    },
    set: function (value) {
      if (typeof value !== "boolean") {
        throw error(
          "InvalidArgument",
          "Workbook.usePrecisionAsDisplayed must be a boolean"
        );
      }
      queueSet(this, "usePrecisionAsDisplayed", value, ensureWorkbookBinding);
    },
  });

  Excel.Workbook.prototype.set = function (source, options) {
    setProperties(
      this,
      source,
      "Workbook",
      ["usePrecisionAsDisplayed"],
      [],
      options
    );
  };

  Excel.Workbook.prototype.toJSON = function () {
    var data = {};
    if (this._loaded.usePrecisionAsDisplayed) {
      data.usePrecisionAsDisplayed = this._usePrecisionAsDisplayed;
    }
    return data;
  };

  // RequestContext owns the same Workbook proxy, so both navigation paths
  // produce one canonical Application object per context.
  Object.defineProperty(Excel.RequestContext.prototype, "application", {
    configurable: true,
    get: function () {
      return this.workbook.application;
    },
  });

  Excel.Application = Application;
  Excel.IterativeCalculation = IterativeCalculation;
})(globalThis);

File: compute/officejs/src/application.rs
//! Office.js application and workbook calculation translation.
//!
//! The JavaScript adapter owns request-context proxy identity.  This module
//! owns the corresponding workbook-backed references and translates the
//! calculation operations into compute-api calls.  In particular, calculation
//! is never implemented by a proxy-side flag: `Application.calculate` must
//! reach the engine's recalculate/rebuild paths.

use std::collections::HashMap;
use std::sync::Arc;

use compute_api::{ComputeApiError, Workbook};
use serde_json::{Map, Value};
use value_types::FiniteF64;

use crate::dispatch::{ExtensionHandler, ExtensionObject, HostDispatchContext};
use crate::host::BatchError;

/// Handler for the application/workbook calculation wire operations emitted
/// by `application.js`.
///
/// The operation names intentionally stay outside the core `Op` enum so the
/// family can be integrated by registering this handler without changing the
/// central host decoder:
///
/// * `applicationBind { id }`
/// * `workbookBind { id }`
/// * `iterativeCalculationBind { id, applicationId }`
/// * `applicationCalculate { id, calculationType }`
pub(crate) struct ApplicationHandler;

impl ExtensionHandler for ApplicationHandler {
    fn can_handle(&self, operation: &str) -> bool {
        matches!(
            operation,
            "applicationBind"
                | "workbookBind"
                | "iterativeCalculationBind"
                | "applicationCalculate"
        )
    }

    fn handle(
        &self,
        operation: &Value,
        context: &mut HostDispatchContext<'_>,
    ) -> Result<bool, BatchError> {
        let name = operation
            .get("op")
            .and_then(Value::as_str)
            .ok_or_else(|| invalid("application operation is missing its op name"))?;

        match name {
            "applicationBind" => {
                let id = required_string(operation, "id")?;
                context.bind_object(&id, Arc::new(ApplicationRef::new(context.workbook())));
                Ok(true)
            }
            "workbookBind" => {
                let id = required_string(operation, "id")?;
                context.bind_object(&id, Arc::new(WorkbookRef::new(context.workbook())));
                Ok(true)
            }
            "iterativeCalculationBind" => {
                let id = required_string(operation, "id")?;
                let application_id = required_string(operation, "applicationId")?;
                let application = context.extension_object::<ApplicationRef>(&application_id)?;
                context.bind_object(&id, Arc::new(application.iterative_calculation()));
                Ok(true)
            }
            "applicationCalculate" => {
                let id = required_string(operation, "id")?;
                let calculation_type = required_string(operation, "calculationType")?;
                let application = context.extension_object::<ApplicationRef>(&id)?;
                application.calculate(calculation_type)?;
                Ok(true)
            }
            _ => Ok(false),
        }
    }
}

/// A workbook-backed `Excel.Application` reference.
#[derive(Clone)]
pub(crate) struct ApplicationRef {
    workbook: Workbook,
}

/// A workbook-backed `Excel.Workbook` reference for calculation settings.
#[derive(Clone)]
pub(crate) struct WorkbookRef {
    workbook: Workbook,
}

/// A workbook-backed `Excel.IterativeCalculation` reference.
#[derive(Clone)]
pub(crate) struct IterativeCalculationRef {
    workbook: Workbook,
}

impl ApplicationRef {
    pub(crate) fn new(workbook: Workbook) -> Self {
        Self { workbook }
    }

    fn iterative_calculation(&self) -> IterativeCalculationRef {
        IterativeCalculationRef {
            workbook: self.workbook.clone(),
        }
    }

    fn calculation_settings(&self) -> Result<Map<String, Value>, BatchError> {
        let settings = self
            .workbook
            .settings()
            .get_workbook_settings()
            .map_err(compute_error)?;
        let value = serde_json::to_value(settings).map_err(encoding_error)?;
        Ok(value
            .get("calculationSettings")
            .and_then(Value::as_object)
            .cloned()
            .unwrap_or_default())
    }

    fn calculation_mode(&self) -> Result<&'static str, BatchError> {
        let calculation = self.calculation_settings()?;
        Ok(match calculation.get("calcMode").and_then(Value::as_str) {
            Some("manual") => "Manual",
            Some("autoNoTable") => "AutomaticExceptTables",
            Some("auto") | None => "Automatic",
            Some(other) => {
                return Err(engine_error(format!(
                    "The engine returned an unknown calculation mode '{other}'"
                )));
            }
        })
    }

    fn calculation_state(&self) -> Result<String, BatchError> {
        self.workbook.calculation_state().map_err(compute_error)
    }

    fn calculate(&self, calculation_type: &str) -> Result<(), BatchError> {
        if !matches!(calculation_type, "Recalculate" | "Full" | "FullRebuild") {
            return Err(invalid(format!(
                "Application.calculate calculationType '{calculation_type}' is invalid"
            )));
        }
        self.workbook
            .calculate(calculation_type)
            .map(|_| ())
            .map_err(compute_error)
    }
}

impl ExtensionObject for ApplicationRef {
    fn object_type(&self) -> &'static str {
        "Application"
    }

    fn load(&self, properties: &[String]) -> Result<HashMap<String, Value>, BatchError> {
        let mut result = HashMap::new();
        for property in properties {
            let value = match property.as_str() {
                "calculationMode" => Value::String(self.calculation_mode()?.to_string()),
                "calculationState" => Value::String(self.calculation_state()?),
                "isNullObject" => Value::Bool(false),
                // Navigation is bound and loaded as its own ClientObject by
                // the JavaScript adapter. It must not be hydrated as a scalar.
                "iterativeCalculation" => continue,
                "activeWindow" | "cultureInfo" | "windows" => {
                    return Err(unsupported(format!(
                        "Application.{property} is not supported by this host"
                    )));
                }
                "calculationEngineVersion"
                | "decimalSeparator"
                | "thousandsSeparator"
                | "useSystemSeparators" => {
                    return Err(unsupported(format!(
                        "Application.{property} is not supported by this host"
                    )));
                }
                other => {
                    return Err(unsupported(format!(
                        "Application.{other} is not supported by this host"
                    )));
                }
            };
            result.insert(property.clone(), value);
        }
        Ok(result)
    }

    fn set(&self, property: &str, value: &Value) -> Result<(), BatchError> {
        match property {
            "calculationMode" => {
                let mode = value.as_str().ok_or_else(|| {
                    invalid("Application.calculationMode must be a string".to_string())
                })?;
                let mode = match mode {
                    "Automatic" => "auto",
                    "AutomaticExceptTables" => "autoNoTable",
                    "Manual" => "manual",
                    other => {
                        return Err(invalid(format!(
                            "Application.calculationMode '{other}' is invalid"
                        )));
                    }
                };
                self.workbook
                    .settings()
                    .set_calculation_mode(mode)
                    .map(|_| ())
                    .map_err(compute_error)
            }
            other => Err(unsupported(format!(
                "Application.{other} is read-only or unsupported"
            ))),
        }
    }
}

impl ExtensionObject for WorkbookRef {
    fn object_type(&self) -> &'static str {
        "Workbook"
    }

    fn load(&self, properties: &[String]) -> Result<HashMap<String, Value>, BatchError> {
        let settings = self
            .workbook
            .settings()
            .get_workbook_settings()
            .map_err(compute_error)?;
        let calculation = settings.calculation_settings.unwrap_or_default();
        let mut result = HashMap::new();
        for property in properties {
            let value = match property.as_str() {
                "usePrecisionAsDisplayed" => Value::Bool(!calculation.full_precision),
                "isNullObject" => Value::Bool(false),
                // The pinned Office.js Workbook declaration has no date1904
                // or date-system property. The engine still retains date1904
                // in WorkbookSettings for formula/date semantics and OOXML
                // round trips; it is deliberately not invented here.
                "date1904" | "dateSystem" | "calculationEngineVersion" => {
                    return Err(unsupported(format!(
                        "Workbook.{property} is not supported by this host"
                    )));
                }
                other => {
                    return Err(unsupported(format!(
                        "Workbook.{other} is not supported by this host"
                    )));
                }
            };
            result.insert(property.clone(), value);
        }
        Ok(result)
    }

    fn set(&self, property: &str, value: &Value) -> Result<(), BatchError> {
        match property {
            "usePrecisionAsDisplayed" => {
                let enabled = value.as_bool().ok_or_else(|| {
                    invalid("Workbook.usePrecisionAsDisplayed must be a boolean".to_string())
                })?;
                let mut settings = self
                    .workbook
                    .settings()
                    .get_workbook_settings()
                    .map_err(compute_error)?;
                let mut calculation = settings.calculation_settings.unwrap_or_default();
                calculation.full_precision = !enabled;
                settings.calculation_settings = Some(calculation);
                self.workbook
                    .settings()
                    .set_workbook_settings(settings)
                    .map(|_| ())
                    .map_err(compute_error)
            }
            other => Err(unsupported(format!(
                "Workbook.{other} is read-only or unsupported"
            ))),
        }
    }
}

impl ExtensionObject for IterativeCalculationRef {
    fn object_type(&self) -> &'static str {
        "IterativeCalculation"
    }

    fn load(&self, properties: &[String]) -> Result<HashMap<String, Value>, BatchError> {
        let settings = self
            .workbook
            .settings()
            .get_workbook_settings()
            .map_err(compute_error)?;
        let calculation = settings.calculation_settings.unwrap_or_default();
        let mut result = HashMap::new();
        for property in properties {
            let value = match property.as_str() {
                "enabled" => Value::Bool(calculation.enable_iterative_calculation),
                "maxChange" => {
                    serde_json::to_value(calculation.max_change).map_err(encoding_error)?
                }
                "maxIteration" => Value::Number(calculation.max_iterations.into()),
                "isNullObject" => Value::Bool(false),
                other => {
                    return Err(unsupported(format!(
                        "IterativeCalculation.{other} is not supported by this host"
                    )));
                }
            };
            result.insert(property.clone(), value);
        }
        Ok(result)
    }

    fn set(&self, property: &str, value: &Value) -> Result<(), BatchError> {
        match property {
            "enabled" => self
                .workbook
                .settings()
                .set_iterative_calculation(value.as_bool().ok_or_else(|| {
                    invalid("IterativeCalculation.enabled must be a boolean".to_string())
                })?)
                .map(|_| ())
                .map_err(compute_error),
            "maxIteration" => {
                let raw = value.as_u64().ok_or_else(|| {
                    invalid(
                        "IterativeCalculation.maxIteration must be a non-negative integer"
                            .to_string(),
                    )
                })?;
                let value = u32::try_from(raw).map_err(|_| {
                    invalid("IterativeCalculation.maxIteration is out of range".to_string())
                })?;
                self.workbook
                    .settings()
                    .set_max_iterations(value)
                    .map(|_| ())
                    .map_err(compute_error)
            }
            "maxChange" => {
                let value = value.as_f64().ok_or_else(|| {
                    invalid(
                        "IterativeCalculation.maxChange must be a finite non-negative number"
                            .to_string(),
                    )
                })?;
                let value = FiniteF64::new(value)
                    .filter(|value| value.get() >= 0.0)
                    .ok_or_else(|| {
                        invalid(
                            "IterativeCalculation.maxChange must be a finite non-negative number"
                                .to_string(),
                        )
                    })?;
                let mut settings = self
                    .workbook
                    .settings()
                    .get_workbook_settings()
                    .map_err(compute_error)?;
                let mut calculation = settings.calculation_settings.unwrap_or_default();
                calculation.max_change = value;
                settings.calculation_settings = Some(calculation);
                self.workbook
                    .settings()
                    .set_workbook_settings(settings)
                    .map(|_| ())
                    .map_err(compute_error)
            }
            other => Err(unsupported(format!(
                "IterativeCalculation.{other} is read-only or unsupported"
            ))),
        }
    }
}

fn required_string<'a>(operation: &'a Value, field: &str) -> Result<&'a str, BatchError> {
    operation.get(field).and_then(Value::as_str).ok_or_else(|| {
        invalid(format!(
            "application operation field '{field}' must be a string"
        ))
    })
}

fn invalid(message: String) -> BatchError {
    BatchError {
        code: "InvalidArgument",
        message,
    }
}

fn unsupported(message: String) -> BatchError {
    BatchError {
        code: "ApiNotFound",
        message,
    }
}

fn engine_error(message: impl Into<String>) -> BatchError {
    BatchError {
        code: "GeneralException",
        message: message.into(),
    }
}

fn compute_error(error: ComputeApiError) -> BatchError {
    engine_error(error.to_string())
}

fn encoding_error(error: serde_json::Error) -> BatchError {
    engine_error(format!("failed to encode workbook settings: {error}"))
}
